ishop/views.py

```diff
@@ -1,3 +1,179 @@
+        in_refund = Refund.objects.filter(purchase__in=self.get_queryset())
+        context['in_refund'] = in_refund.values_list('purchase__pk', flat=True)
+        context['balance'] = ShopUser.objects.get(id=self.request.user.id).wallet
+        return context
+
+    def post(self, request, *args, **kwargs):
+        purchase = Purchase.objects.get(pk=self.request.POST['pk'])
+        delta = timezone.now() - timezone.timedelta(minutes=self.time_interval_to_refund)
+        if purchase.datetime < delta:
+            msg = 'Your refund time has been expired'
+            messages.error(self.request, msg)
+            return self.get(request, *args, **kwargs)
+
+        _, is_new = Refund.objects.get_or_create(purchase=purchase)
+        if is_new:
+            msg = 'Your refund request has been sent. Wait for approving.'
+            messages.success(self.request, msg)
+        else:
+            msg = 'The refund for current purchase already exists. Wait for approving.'
+            messages.error(self.request, msg)
+
+        return self.get(request, *args, **kwargs)
+
+
+class AdminRefundView(SuperUserRequiredMixin, ListView):
+    model = Refund
+    paginate_by = 10
+    template_name = 'admin_refunds.html'
+    extra_context = {'title': 'Admin: Purchases to refund'}
+
+    def post(self, request, *args, **kwargs):
+        pk = self.request.POST['pk']
+        approval = self.request.POST['approval']
+
+        if approval == 'decline':
+            self.model.objects.get(pk=pk).delete()
+        else:
+            refund = self.model.objects.get(pk=pk)
+            user = refund.purchase.customer
+            good = refund.purchase.good
+            quantity = refund.purchase.quantity
+            price = refund.purchase.price
+
+            user.wallet += price * quantity
+            user.save()
+
+            good.in_stock += quantity
+            good.save()
+
+            self.model.objects.get(pk=pk).delete()
+            refund.purchase.delete()
+
+        return self.get(request, *args, **kwargs)
+
+
+class AdminGoodsView(SuperUserRequiredMixin, ListView):
+    model = Good
+    paginate_by = 20
+    template_name = 'admin_goods.html'
+    extra_context = {'title': 'Admin: Goods'}
+
+
+class AdminGoodEditView(SuperUserRequiredMixin, UpdateView):
+    model = Good
+    fields = ['title', 'description', 'price', 'image', 'in_stock']
+    template_name = 'admin_good_edit.html'
+    extra_context = {'title': 'Admin: edit'}
+    success_url = '/admin-goods'
+
+
+class AdminGoodAddView(SuperUserRequiredMixin, CreateView):
+    model = Good
+    template_name = 'admin_good_add.html'
+    fields = '__all__'
+    success_url = '/admin-goods'
+    extra_context = {'title': 'Admin: add new good'}
+
+
+class Login(LoginView):
+    success_url = '/'
+    template_name = 'login.html'
+
+    def get_success_url(self):
+        return self.success_url
+
+
+class Register(CreateView):
+    User = get_user_model()
```

Comment:
Why 2 strings?? cannot understand logic

Reply:
It is advised on the Internet - to get the model from the settings, not to write directly
the reason: when there are on site few apps, it helps not to mix up sources of the User model

here I deleted it and wrote it directly